### PR TITLE
Support for ISO 19115 Part 3 XML

### PR DESCRIPTION
**Pull request by vjf**
# Overview - Support for ISO 19115 Part 3 XML

This PR contains a new profile for pycsw, called 'iso19115p3'.

* The profile is in 'pycsw\pycsw\plugins\profiles\iso19115p3', this includes a new outputschema
* There is a new file 'mdb/py' in 'pycsw/pycsw/core' that contains classes used to parse the XML.
* There is a unit test for 'mdb.py' in 'pycsw\tests\unittests\test_mdb.py'
* There are function tests in 'pycsw\tests\functionaltests\suites\mdb'

# Related Issue / Discussion

# Additional Information

# Contributions and Licensing

(as per https://github.com/geopython/pycsw/blob/master/CONTRIBUTING.rst#contributions-and-licensing)

- [X] I'd like to contribute ISO 19115 Part 3 support for pycsw. I confirm that my contributions to pycsw will be compatible with the pycsw license guidelines at the time of contribution.
- [X] I have already previously agreed to the pycsw Contributions and Licensing Guidelines

### CODE REVIEW

**Comment by kalxas** [comment]
Thank you @vjf for this contribution, we will need some time to review.
My first comment is: why not add the parsing capability in OWSLib first?

**Comment by vjf** [comment]
@kalxas Well, amending all the geopython packages to understand ISO 19115-3 XML is quite a large job.
It has to be broken up into parts.

There is a demand from the user community for this in pycsw, ckan etc. so I thought this was the best place to start.

I see integrating ISO 19115-3 XML support into OWSLib as the next stage.
I haven't looked into how much work this would require.

If you like, I can make a PR for OWSLib and insert the new parser and unit tests. It might need "OWSLib >=  XX.XX"  in pycsw's requirements file and will require some release coordination timing between OWSLib and pycsw.

Alternatively we can have pycsw use the local copy of the parser now, then once OWSLib has its own parser, pycsw can preferentially load it. We can remove the local parser from pycsw later on.

**Comment by tomkralidis** [comment]
Great work @vjf!  Agree with @kalxas, for maximum reuse and benefit, the MDB parser should be implemented in OWSLib proper (in `owslib/iso_3.py`), at which point we can cut an OWSLib release that will be used by pycsw.  At that point we can continue the review.

**Comment by vjf** [comment]
@tomkralidis @kalxas  OK, I will migrate mdb.py to OWSLib.

**Comment by kalxas** [comment]
@vjf now that the OWSLib pull request has been merged, can you please make the changes here to align?

**Comment by vjf** [comment]
@tomkralidis @kalxas 

Almost there...

1. I have added a fix to OWSLib https://github.com/geopython/OWSLib/pull/924
2. The most recent OWSLib release will not work with this PR. You'll need to do an OWSLib release, right?

**Comment by tomkralidis** [comment]
@vjf FYI OWSLib 0.31.0 has now been released.

**Comment by kalxas** [comment]
CI is still failing I see

**Comment by vjf** [comment]
@kalxas @tomkralidis All fixed. Thanks.

**Comment by kalxas** [comment]
@vjf thank you.
We will review and come back with comments

**Comment by giohappy** [comment]
We're planning to upgrade PyCSW for GeoNode and one of the compelling reasons is the support for ISO 19115 Part 3.
Do you plan to merge this before the release of PyCSW 3?

I take the opportunity to ask if you have a timeline for the release of PyCSW 3.

**Comment by tomkralidis** [comment]
@kalxas let's review during this week's OGC Sprint.

File: tests/functionaltests/suites/cite/data/cite.db
Comment (tomkralidis) [inline tests/functionaltests/suites/cite/data/cite.db:1]: `tests/functionaltests/suites/cite/data/cite.db` should not have any updates (only for CITE tests with official CITE test data).  If test data is needed, suggest another SQLite database file in `tests/functionaltests/suites/mdb/data/cite.dbtests/functionaltests/suites/cite/data`.

File: pycsw/core/config.py
Comment (tomkralidis) [inline pycsw/core/config.py:139]: We need to validate these queryables (see other comments in this review).

File: pycsw/core/config.py
Comment (tomkralidis) [inline pycsw/core/config.py:176]: Same here, we need to validate this queryable (see other comments in this review).

File: pycsw/core/repository.py
Comment (tomkralidis) [inline pycsw/core/repository.py:820]: The constitutes a model change (and DB migration for existing installations).  Just want to confirm that `funder` is a valuable queryable and column to have in the database model.

File: pycsw/core/repository.py
Comment (tomkralidis) [inline pycsw/core/repository.py:844]: These constitutes a model change (and DB migration for existing installations).  Just want to confirm that `vert_extent_min` and `vert_extent_max` are valuable queryables and columns to have in the database model.

File: setup.py
Comment (tomkralidis) [inline setup.py:72]: Remove trailing comma not needed.

**Review by tomkralidis — CHANGES_REQUESTED** [review]
Thank you for your patience @vjf -- an excellent piece of work here!  See change requests.

Additional comment: the directory `tests/functionaltests/suites/mdb` is better renamed to `tests/functionaltests/suites/iso19115p3`.

File: tests/functionaltests/suites/cite/data/cite.db
Comment (vjf) [reply tests/functionaltests/suites/cite/data/cite.db:1]: @tomkralidis I have removed it from the pull request

File: setup.py
Comment (vjf) [reply setup.py]: @tomkralidis Fixed.

File: pycsw/core/repository.py
Comment (vjf) [reply pycsw/core/repository.py:821]: @tomkralidis We are under some pressure to recognise sources of funding in all our metadata. This helps us justify to our funders that their money does have valuable and tangible outcomes.   But it is your call in the end.

File: pycsw/core/config.py
Comment (vjf) [reply pycsw/core/config.py:139]: @tomkralidis I have colleagues who work in ocean biosciences for whom this is a fundamental requirement. For myself, I require it to represent underground geology/geophysics datasets.

File: pycsw/core/config.py
Comment (vjf) [reply pycsw/core/config.py:176]: @tomkralidis We are under some pressure to recognise sources of funding in all our metadata. This helps us justify to our funders that their money does have valuable and tangible outcomes. But it is your call in the end.

File: pycsw/core/repository.py
Comment (vjf) [reply pycsw/core/repository.py:845]: @tomkralidis See previous comments.

**Comment by vjf** [comment]
@tomkralidis @kalxas All fixed.

File: pycsw/core/config.py
Comment (tomkralidis) [reply pycsw/core/config.py:176]: @vjf in discussion with @kalxas, we do not store role codes as queryables currently.  We are fine with adding vertical extents.

**Comment by tomkralidis** [comment]
@vjf sorry for the delay; it seems I added some final comments to this PR but they were not submitted.

We are fine with adding the extra columns for vertical extent (min and max), but prefer that that the funder column be managed outside of pycsw as an extra column.  Once this is updated, I'll merge this PR -- great work again!

**Comment by vjf** [comment]
@tomkralidis @kalxas Removed 'Funder' as requested. Thanks.

**Review by tomkralidis — APPROVED** [review]

**Comment by tomkralidis** [comment]
Thanks @vjf, merged.

**Pull request merged**
